presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/ShardCleaner.java

```diff
@@ -327,35 +332,79 @@ synchronized void cleanLocalShards()
     @VisibleForTesting
     void cleanBackupShards()
     {
+        Set<UUID> processing = newConcurrentHashSet();
+        BlockingQueue<UUID> completed = new LinkedBlockingQueue<>();
+        long start = System.nanoTime();
+        boolean fill = true;
+
         while (!Thread.currentThread().isInterrupted()) {
-            List<UUID> uuids = dao.getCleanableShardsBatch(maxTimestamp(backupCleanTime));
-            if (uuids.isEmpty()) {
+            // get a new batch if any completed and we are under the batch size
+            Set<UUID> uuids = ImmutableSet.of();
+            if (fill && (processing.size() < 1000)) {
+                uuids = dao.getCleanableShardsBatch(maxTimestamp(backupCleanTime));
+                fill = false;
+            }
+            if (uuids.isEmpty() && processing.isEmpty()) {
                 break;
             }
 
-            long start = System.nanoTime();
-            executeDeletes(uuids);
-            dao.deleteCleanedShards(uuids);
-            backupShardsCleaned.update(uuids.size());
-            log.info("Cleaned %s backup shards in %s", uuids.size(), nanosSince(start));
+            // skip any that are already processing and mark remaining as processing
+            uuids = ImmutableSet.copyOf(difference(uuids, processing));
+            processing.addAll(uuids);
+
+            // execute deletes
+            for (UUID uuid : uuids) {
+                runAsync(() -> backupStore.get().deleteShard(uuid), backupExecutor)
+                        .thenAccept(v -> completed.add(uuid))
+                        .whenComplete((v, e) -> {
+                            if (e != null) {
+                                log.error(e, "Error cleaning backup shard: %s", uuid);
```

Comment:
So if we have a problem with more thank 1k shards, this code could be stuck without progress.  Of course there would be a ton of log messages to let us know.